### PR TITLE
Support for files and cover image

### PR DESCRIPTION
**Pull request by meinaart**
I have added support for cover image and other types of assets. It was on the todo list and I needed it, I hope you like it. I tried to stay within your code style.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Change Log
```
Comment (iarna) [inline CHANGELOG.md:1]: I appreciate this addition but it should be in a separate commit.

File: CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+All notable changes to this project will be documented in this file.
+
+## [1.0.5] - 2016-09-23
```
Comment (iarna) [inline CHANGELOG.md]: This should actually be a feature bump (1.1.0)

File: README.md [hunk 3]
```diff
@@ -35,8 +36,10 @@ All of the options can be set after object creation with obvious setters:
 ### epub.setModified(modified)
 ### epub.setPublished(published)
 ### epub.setSource(source)
-### epub.setLanguage(language)
 ### epub.setDescription(description)
+### epub.setLanguage(language)
+
+Identifies the language used in the book content. The content has to comply with [RFC 3066](http://www.ietf.org/rfc/rfc3066.txt). [List of language codes](http://www.loc.gov/standards/iso639-2/php/code_list.php).
```
Comment (iarna) [inline README.md]: _language_ was already described on line 28, but I'm happy to merge this into that.

File: README.md [hunk 4]
```diff
@@ -46,18 +49,31 @@ can pipe into it or write to it and call `.end()` when you're done.
...
 
-* **chapterName** - The name of the chapter in the index.
-* **content** - The HTML content of this chapter.  This will be passed
+#### Adding a chapter
+The object must have the following keys:
+* **`chapterName`** - The name of the chapter in the index.
+* **`content`** - *(optional)* The HTML content of this chapter.  This will be passed
   through `htmltidy` in order to make it valid XHTML.
-* **index** - *(optional)* Where the chapter should show up in the index. These numbers
+* **`sourceFileName`** - *(optional)* Path to the source HTML file with chapter
```
Comment (iarna) [inline README.md]: I'm not super keen having streampub ever read from disk itself. How would you feel about allowing `content` to be a stream in addition to a string?

Conveniently html-tidy has a streaming mode. Annoyingly zip-stream does not, so we'd have to concat it at that point, but oh well. Something to patch later. ;)

File: README.md [hunk 5]
```diff
@@ -46,18 +49,31 @@ can pipe into it or write to it and call `.end()` when you're done.
...
 
+(`sourceFilename` or `content` should be set)
+
+#### Adding other type of assets
+You can add other types of files to the epub. For example images or stylesheets.
+
+The object must have the following keys:
+* **fileName** - The filename to use *inside* the epub.
+* **sourceFileName** - Path to the source file.
+
```
Comment (iarna) [inline README.md]: I saw that you added support for explicitly passing in mime types. That should be documented here and above.

Same thing applies here, I'd prefer that there not be a `sourceFileName` and instead be a `content` field that can be a buffer, string or stream.

File: README.md [hunk 6]
```diff
@@ -73,6 +89,24 @@ This is the usual stream write function. The object must have the following keys
...
+var fs = require('fs')
+
+var epub = new Streampub({title: 'My Example'})
+epub.setAuthor('Example author')
+epub.pipe(fs.createWriteStream('example.epub'), {end: false})
+epub.write({id: 'cover-image', sourceFileName: 'image.jpg'})
+epub.write({chapterName: 'Chapter 1', content: '<h1>Chapter 1</h1><b>doc content</b>'})
+epub.write({chapterName: 'Chapter 2', content: '<h1>Chapter 2</h1><b>doc content</b>'})
+epub.end()
+```
```
Comment (iarna) [inline README.md:139]: I love examples!

File: index.js [hunk 7]
```diff
@@ -68,20 +90,81 @@ function Chapter (index, chapterName, fileName, content) {
   return {index: index, chapterName: chapterName, fileName: fileName, content: content}
 }
 
+Streampub.prototype._generateCover = function(async) {
```
Comment (iarna) [inline index.js]: Sooo, two things…

`async` is likely to be a keyword [Real Soon Now](https://github.com/tc39/ecmascript-asyncawait), so it shouldn't be a variable name.

And I'd would prefer this not be functionality that's controlled by a flag. I'd prefer two functions, perhaps with one calling the other.

File: index.js [hunk 8]
```diff
@@ -68,20 +90,81 @@ function Chapter (index, chapterName, fileName, content) {
...
+  var self = this
+  var title = self.meta.title || 'Cover'
+  function execute(callback) {
+    self.write({
```
Comment (iarna) [inline index.js]: As you've discovered ("write after end") calling "write" from the guts of a stream implementation is fraught and you shouldn't really do it.

If you want to do this calling `self._transform` would be way safer.

File: index.js [hunk 9]
```diff
@@ -46,13 +54,27 @@ function Streampub (opts) {
 }
 inherits(Streampub, Transform)
 
+// Theoretically this could be done in _flush, but somehow this generates an error
+// this is a workaround for that error ('no write after end')
+Streampub.prototype._end = Streampub.prototype.end;
```
Comment (iarna) [inline index.js]: I discuss this a bit more below, but this should definitely be done in flush. If `_generateCover` uses `self._transform` instead of `self.write` then it should work fine. No need to wait on the callback even.

File: index.js [hunk 10]
```diff
@@ -68,20 +90,81 @@ function Chapter (index, chapterName, fileName, content) {
...
+    if(data.chapterName) {
+      self.chapters[index] = {index: index, chapterName: data.chapterName, fileName: data.fileName}
+    }
+    self.files.push({chapterName: data.chapterName, fileName: data.fileName, mime: data.mime, id: data.id || 'file' + id})
     self.header.then(function () {
-      return self.zip.entry(html, {name: 'OEBPS/' + fileName})
-    }).finally(function () {
-      done()
+      return self.zip.entry(content, {name: 'OEBPS/' + data.fileName})
+    }).finally(done)
```
Comment (iarna) [inline index.js]: good catch

File: package.json [hunk 11]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "streampub",
-  "version": "1.0.4",
+  "version": "1.0.5",
```
Comment (iarna) [inline package.json]: I do version bumps separate from changes (with `npm version`)

**Review by iarna — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 2 above)
Comment (iarna) [reply CHANGELOG.md]: well, 1.2.0 now =)

File: CHANGELOG.md (hunk 1 above)
Comment (meinaart) [reply CHANGELOG.md:1]: Sure no problem. I can add it in a separate commit. It does make history a bit more readable.

File: README.md (hunk 3 above)
Comment (meinaart) [reply README.md]: Good point, I did notice some description on language after I wrote this. I will try to merge the 2.

File: CHANGELOG.md (hunk 2 above)
Comment (meinaart) [reply CHANGELOG.md]: You are right, I should have followed semver a bit better ;). I will try to bring my PR up to date with your latest changes and follow semver numbering.

File: README.md (hunk 4 above)
Comment (meinaart) [reply README.md]: Yeah I agree with you, it deviates too much from the rest of the module. I think it's nice to have a module that works purely with streams.

File: README.md (hunk 5 above)
Comment (meinaart) [reply README.md]: Yeah I agree with you once again. But I don't really like this approach normally. I would rather supply an in between object here. For example a `Chapter`, `Asset` or `CoverImage` object with some required fields.

Not sure on syntax yet but something like this might work:

```
var chapter = new Chapter(title, content, index);
chapter.tidyHtml = true;
streampub.write(chapter);

var asset = new Asset(filename, content);
streampub.write(asset);

var coverImage = new CoverImage(content);
coverImage.autoGenerateHtml = true;
streampub.write(coverImage);
```

This approach would allow us a clean separation of logic (like auto generation of cover HTML, or tidying content).

File: README.md (hunk 6 above)
Comment (meinaart) [reply README.md:139]: Yeah me too! :) Maybe a separate example section in the `README.md` would be nice.

File: index.js (hunk 9 above)
Comment (meinaart) [reply index.js]: Yeah I tried that, but somehow it did not work correctly. If I remember correctly it did seem to work but the `cover.html` did not end up in the zip file.

I should look at it again. It didn't feel right doing it like this.

File: index.js (hunk 7 above)
Comment (meinaart) [reply index.js]: It's actually sort of a leftover of a few things I tried because of "write after end" thing I ran into. I will clean it up. Thanks for pointing that `async` thing out.

File: index.js (hunk 8 above)
Comment (meinaart) [reply index.js]: See my comment above, I did not get it to work as I wanted. But I will look at it again because I agree that it seems hacky like this.

File: index.js (hunk 10 above)
Comment (meinaart) [reply index.js]: Thanks.

File: package.json (hunk 11 above)
Comment (meinaart) [reply package.json]: Alright, good to know. I will refrain from bumping versions then :).

File: CHANGELOG.md (hunk 1 above)
Comment (meinaart) [reply CHANGELOG.md:1]: I put it in a separate commit.

File: CHANGELOG.md (hunk 2 above)
Comment (meinaart) [reply CHANGELOG.md]: It's updated now to 1.2.0 in changelog, not in package.json

**Comment by meinaart** [comment]
I did something bad :P. I force pushed some changes based on your latest version. I can create a new pull request if that makes it a bit more readable or you can review as is.

I think I processed all your feedback.

**Comment by iarna** [comment]
I like force pushes on PRs! Makes it all way more readable. Plus with the new review interface, previous comments don't go away!

**Review by iarna — APPROVED** [review]

**Comment by iarna** [comment]
This is all merged and a new version is published. (Uhhm, I screwed up and squashed them together into the changelog commit. And didn't notice till _after_ I'd published and pushed. So the git history is a little confusing. =D)

**Comment by iarna** [comment]
For the record, that's b0a77b4.
